### PR TITLE
Implement `GroupHash` unmerging.

### PR DESCRIPTION
**Pull request by tkaemming**
This allows disassociating a `GroupHash` record from a group. This causes a new group to be created, and denormalized statistics (time series and tag data, among others) are backfilled for both the source and destination groups.

This implementation does have a few caveats:

- if sampling is enabled, statistics will be incorrect (by a factor of the sampling ratio, so assuming a uniform distribution, the proportions of the statistics should still be representative of the inputs),
- this cannot handle checksums provided as part of the data payload, since they are deleted from the payload on ingest,
- this assumes no modifications to the fingerprinting algorithm, since calculated fingerprint values are never associated with the event when using default fingerprints.

### CODE REVIEW

**Comment by tkaemming** [comment]
There are a few things that need to happen before this is ready to be merged (mostly around performance, and validating a few questions that are noted inline), but I think the structure of this patch is unlikely to change in any significant way unless there is something I totally overlooked.

File: src/sentry/tasks/unmerge.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+                ),
+            ),
+        ),
+    )
+
+    for event in events:
+        counters[event.datetime][tsdb.models.group][event.group_id] += 1
+
+        user = event.data.get('sentry.interfaces.User')
+        if user:
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: This block isn't covered in tests, I'll get to this.

File: tests/sentry/tasks/test_unmerge.py [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+            'active_at': now - timedelta(hours=2),
+            'first_seen': now - timedelta(hours=2),
+            'platform': 'java',
+            'score': ScoreClause.calculate(3, now),
+            'logger': 'java',
+            'times_seen': 3,
+            'first_release': None,
+        }
+
+    def test_unmerge(self):
```
Comment (tkaemming) [inline tests/sentry/tasks/test_unmerge.py]: I hate basically everything about this test case.

File: src/sentry/tasks/unmerge.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+initial_fields = {
+    'culprit': lambda event: generate_culprit(
+        event.data,
+        event.platform,
+    ),
+    'data': lambda event: {
+        'last_received': event.data.get('received') or float(event.datetime.strftime('%s')),
+        'type': event.data['type'],
+        'metadata': event.data['metadata'],
+    },
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: I'm not 100% sure this data has existed forever, and there's also no explicit schema that requires it, but without it, a lot of code would break … so I'm just sort of assuming it exists … ?

File: src/sentry/tasks/unmerge.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+        # XXX: There is a race condition here between the (wall clock) time
+        # that the migration is started by the user and when we actually
+        # get to this block where the new destination is created and we've
+        # moved the ``GroupHash`` so that events start being associated
+        # with it. During this gap, there could have been additional events
+        # ingested, and if we want to handle this, we'd need to record the
+        # highest event ID we've seen at the beginning of the migration,
+        # then scan all events greater than that ID and migrate the ones
+        # where necessary. (This still isn't even guaranteed to catch all
+        # of the events due to processing latency, but it's a better shot.)
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: Unsure how much we care about this since there are so many other ways this could break while only partially completed.

File: src/sentry/tasks/unmerge.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+            short_id=project.next_short_id(),
+            **get_group_creation_attributes(events)
+        )
+
+        destination_id = destination.id
+
+        # Move the group hashes to the destination.
+        # TODO: What happens if this ``GroupHash`` has already been
+        # migrated somewhere else? Right now, this just assumes we have
+        # exclusive access to it (which is not a safe assumption.)
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: Was thinking we'd probably add a status column to `GroupHash` to allow best-effort exclusive access here.

File: src/sentry/tasks/unmerge.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+
+def update_tag_value_counts(id_list):
+    instances = GroupTagKey.objects.filter(group_id__in=id_list)
+    for instance in instances:
+        instance.update(
+            values_seen=GroupTagValue.objects.filter(
+                project_id=instance.project_id,
+                group_id=instance.group_id,
+                key=instance.key,
+            ).count(),
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: I'm hoping this is executed as a subquery to minimize data races but I need to double check.

File: src/sentry/tasks/unmerge.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+
+    for event in events:
+        event.group = destination
+
+    EventTag.objects.filter(
+        project_id=project.id,
+        event_id__in=event_id_set,
+    ).update(group_id=destination_id)
+
+    event_event_id_set = set(event.event_id for event in events)
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: This line is a personal favorite.

File: src/sentry/tasks/unmerge.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+
+
+def merge_mappings(values):
+    result = {}
+    for value in values:
+        result.update(value)
+    return result
+
+
+initial_fields = {
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: These are all the fields that have last-write-wins semantics in `EventManager`, so they're written first and never updated. If other events are ingested while this backfill is occurring, the new events will overwrite these values which is intended.

I didn't make any editorial decisions about what fields to include here rather than as part of the backfill fields, so if you disagree with some of those choices, that's a discussion to have outside of the scope of this patch.

File: src/sentry/tasks/unmerge.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+
+        # TODO: Double check this!
+        if not release:
+            continue
+
+        # XXX: It's not really clear what the canonical source is for
+        # environment between the tag and the data attribute, but I'm going
+        # with data attribute for now. Right now it seems like they are
+        # intended to both be present and the same value, but I'm not really
+        # sure that has always been the case for existing values.
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: This is another case where I'm not sure what the implications of assuming this data has existed forever are.

File: src/sentry/tasks/unmerge.py [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+        lambda: defaultdict(
+            lambda: defaultdict(int),
+        ),
+    )
+
+    frequencies = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(
+                lambda: defaultdict(
+                    int,
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: Not sure why I made a stylistic distinction here compared to the lines above. 🤔

File: src/sentry/tasks/unmerge.py [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+                EventUser(
+                    project=project,
+                    ident=user.get('id'),
+                    email=user.get('email'),
+                    username=user.get('username'),
+                    ip_address=user.get('ip_address'),
+                ).tag_value
+            )
+
+        environment = Environment.objects.get(
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: This should also be cached.

File: src/sentry/tasks/unmerge.py (hunk 6 above)
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]:
```
LOG:  duration: 0.136 ms  statement: SELECT COUNT(*) FROM "sentry_messagefiltervalue" WHERE ("sentry_messagefiltervalue"."project_id" = 3  AND "sentry_messagefiltervalue"."group_id" = 9  AND "sentry_messagefiltervalue"."key" = 'url' )
LOG:  duration: 0.010 ms  statement: BEGIN
LOG:  duration: 0.136 ms  statement: UPDATE "sentry_grouptagkey" SET "values_seen" = 1 WHERE "sentry_grouptagkey"."id" = 68 
LOG:  duration: 0.162 ms  statement: COMMIT
```
🙄

**Comment by JTCunning** [comment]
Did a bunch of re-arranging of workflow because I think we need platform keyed into this functionality.

File: src/sentry/tasks/unmerge.py [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+def collect_tsdb_data(project, events):
+    counters = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(int),
+        ),
+    )
+
+    sets = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(int),
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: This is wrong, I'll fix when I add coverage below to the user tracking path. ("Static type systems don't catch bugs, they said…")

**Review by mitsuhiko — COMMENTED** [review]
I mean this seems relatively okay but I have not spent too much time contemplating about what this means for ops.

I do want to make a second pass over it later. One thing I do want to mention though is that this seems to duplicate some logic and I fear we might forget to update this later.

File: src/sentry/tasks/unmerge.py [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+    frequencies = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(
+                lambda: defaultdict(int),
+            ),
+        ),
+    )
+
+    for event in events:
+        counters[event.datetime][tsdb.models.group][event.group_id] += 1
```
Comment (LewisJEllis) [inline src/sentry/tasks/unmerge.py]: at first it wasn't clear why we have the `tsdb.models.group` bit in here since it's the only 2nd-level key used (same sitch for `sets` and `tsdb.models.users_affected_by_group`), but I guess since frequencies uses multiple models and `tsdb.incr` and `tsdb.record` obviously want the model ID, this is a more general/flexible pattern to separate the collection from the writing. once i got that i realized this is actually pretty nice because it lets this method worry about the data in particular and `repair_tsdb_data` just care about *doing* all the writes, agnostic to the models involved.

tldr "was almost confusing at first but no it makes perfect sense"

**Review by LewisJEllis — APPROVED** [review]
Makes sense overall, structured well, looks good. 

Definitely lots of "todos" and "make sure..."s though, none super worrying but maybe take a quick pass over those before merging in case any have been resolved? or make sure they're planned for another look in the future

File: src/sentry/tasks/unmerge.py (hunk 4 above)
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: 🙃

File: src/sentry/tasks/unmerge.py [hunk 14]
```diff
@@ -0,0 +1,544 @@
...
+def get_fingerprint(event):
+    # TODO: This *might* need to be protected from an IndexError?
+    primary_hash = get_hashes_for_event(event)[0]
+    return md5_from_hash(primary_hash)
+
+
+def migrate_events(caches, project, source_id, destination_id, fingerprints, events, actor_id):
+    # XXX: This is only actually able to create a destination group and migrate
+    # the group hashes if there are events that can be migrated. How do we
+    # handle this if there aren't any events? We can't create a group (there
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: > would it ever actually happen that there aren't any events? seems like if we ever do this operation there must be events to migrate, can we not make/rely on that guarantee?

Nothing enforces it, so it's theoretically possible and I'd like to plan for it (or at least note that it's possible, like here.) A short aside: in my opinion it's a good idea to never assume anything is going to work as it "should" unless a more reliable piece of technology enforces that assumed constraint. This is why I advocate so much for maintaining referential integrity through things like foreign key constraints. (I'm also aware that this isn't necessarily a shared opinion of the team.)

The only ways that I can think of this would happen is (note that this list isn't exhaustive, it's just what I could come up with off of the top of my head):

1. All of the events with a specific hash are deleted via cleanup, or some other means. (I'm not even sure if there is an API endpoint for this, off of the top of my head?)
2. [A "secondary" fingerprint could be assigned to a group if it returns multiple hashes](https://github.com/getsentry/sentry/blob/fe03d388e52fe7c73330857c59b0f4f3fd6104ac/src/sentry/event_manager.py#L836-L849) (e.g. application hash and system hash), and since this task only unmerges the first returned ("primary") hash for event (see L136 in this file), if you tried to unmerge the "secondary" hash, nothing would happen.

I don't think it's likely, though, except maybe in that second case. Plus, what would even be migrated, other than just the `GroupHash` record itself? I think we'll probably want to improve the UX here in the future, but I don't have any good ideas off of the top of my head.

File: src/sentry/tasks/unmerge.py [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+
+    for event in events:
+        tags = results.setdefault(event.group_id, {})
+
+        for key, value in event.get_tags():
+            values = tags.setdefault(key, {})
+
+            if value in values:
+                times_seen, first_seen, last_seen = values[value]
+                values[value] = (times_seen + 1, event.datetime, last_seen)
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: It's not strong at all (in fact it's likely wrong), but this was intentionally written this way to emulate the existing behavior. (There are more details over at GH-5289 if you're interested.)

File: src/sentry/tasks/unmerge.py (hunk 6 above)
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: > I sorta wondered if we could instead just update this as we go in repair_tag_data, and we could, but it would mean a lot more writes in high-tag-value-cardinality cases, so almost definitely not worth it.

That was my first thought too, but that introduces two issues:

1. tracking high cardinality values (as you mentioned) could become expensive,
2. it also increases the potential for inconsistency here due to data races because any new tag keys that were associated with this issue after the `GroupHash` was moved will be ignored.

>  How hard is it to get this to run as a subquery? I guess that's probably a django ORM question?

Not sure, but also a little hesitant to add the complexity here of dropping to raw SQL just for the sake of optimizing this into a single transaction. (I might take a look though in case I'm wrong.)

File: src/sentry/tasks/unmerge.py [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+    sets = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(set),
+        ),
+    )
+
+    frequencies = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(
+                lambda: defaultdict(int),
```
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: 😎

File: src/sentry/tasks/unmerge.py (hunk 13 above)
Comment (tkaemming) [inline src/sentry/tasks/unmerge.py]: Yep, this was my goal. It's a little weird to read at first (the fact that the TSDB API itself is sort of weird/limiting doesn't help either) but this is a hell of a lot easier to test when the query construction and execution is decoupled, in my opinion.
